### PR TITLE
[156] Downsample photos with too big a resolution

### PR DESCRIPTION
**Pull request by jbors**
Fix for https://github.com/theaiscope/GDD-app/issues/21

Downsample photo's until they are beneath the max canvas size

### CODE REVIEW

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            decodeSampledBitmapAndCache(targetFile, 20, 20, cacheDir)
+        }
+
+        assertTrue(cacheDir.isDirectory)
+        val cachedFiles = cacheDir.listFiles()
+
+        assertTrue(cachedFiles.any { file -> file.name.equals("test_20x20.bmp", true) })
+        assertTrue(b.height < 50)
+        assertTrue(b.width < 50)
+        assertTrue(b.height > 20)
```
Comment (eduardb) [inline app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt:43]: Should we assert equality instead? Or do you think that this would couple the test too much to the implementation details? Either way, 20 would be a valid dimension as well. so this should be `>=`, no?

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    private val targetContext = InstrumentationRegistry.getInstrumentation().targetContext
+    private val cacheDir = targetContext.cacheDir
+    private lateinit var targetFile: File
+
+    @Before
+    fun setup() {
+        targetFile = File(targetContext.filesDir, "test.bmp");
+        targetFile.createNewFile()
+
+        val stream = testContext.assets.open("w3c_home.bmp");
```
Comment (eduardb) [inline app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt:26]: In `MaskActivityTest.kt` there's a `getAssetStream` method that does this, and even has a failback. I think that it could be made an extension function on `Context`, and be reused here and in `BitmapReaderTest` as well, what do you think?

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.io.File
+import java.io.InputStream
+
+@RunWith(AndroidJUnit4::class)
+class SampleImagesAdapterTest {
```
Comment (eduardb) [inline app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt:15]: The tests in this class don't really have to do anything with the adapter itself, but with some helper method that happens to be in the same file as the adapter. Thus, I am thinking, maybe it would make sense to move that method to its own file? Honestly, I wouldn't mind having it in the `BitmapReader` class, but whatever you prefer :)

File: app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt [hunk 4]
```diff
@@ -86,73 +89,47 @@ private class ImageViewHolder(
         itemView.setOnClickListener { onImageClicked(imageFile, maskFile) }
     }
 
-    fun bind(image: File, mask: File) {
+    fun bind(image: File, mask: File, cacheDir: File) {
```
Comment (eduardb) [inline app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt:92]: `cacheDir` is not used, did you mean using it below instead of `itemView.context.cacheDir`? Otherwise, it can be removed.

File: app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt [hunk 5]
```diff
@@ -86,73 +89,47 @@ private class ImageViewHolder(
...
-        // Calculate inSampleSize
-        inSampleSize = calculateInSampleSize(this, reqWidth, reqHeight)
-
-        // Decode bitmap with inSampleSize set
-        inJustDecodeBounds = false
-
-        val bitmap = BitmapFactory.decodeFile(image.absolutePath, this)
+    val bitmap = BitmapReader.decodeSampledBitmapFromResource(
+        image,
+        false,
```
Comment (eduardb) [inline app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt:125]: Would you use a named argument here, so it's clear what does the `false` stands for, please?

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 3 above)
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt:15]: I had it in BitmapReader initially but moved it back for a reason that now does not exist anymore 😆 

Probs best to move it back

File: app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt (hunk 5 above)
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt]: Had to change the order since it's not allowed to mix positional & named arguments, but done

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 3 above)
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Moved

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 1 above)
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Well I don't want to pin myself down to a certain number as a result since that's not explicitly expected. You're right on the equality though, adjusted (even though it does not occur with this specific implementation ;))

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 2 above)
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Done. Also moved the `ParcelableExt` to a new `test.extensions` package

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+        assertTrue(cachedFiles.any { file -> file.name.equals("test_20x20.bmp", true) })
+        assertTrue(b.height < 50)
+        assertTrue(b.width < 50)
+        assertTrue(b.height > 20)
+        assertTrue(b.width > 20)
+    }
+
+    @Test
+    fun shouldNotWriteToCacheTwice() {
```
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: So apparently the `lastmodified` is in seconds. So the test runs in <1 second and you get the same value on `lastmodified`

I added a Thread.sleep(2000) between the first and second calls, and then it will fail if you do not return the cached file. But that's a bit, ewww... Suggestions?

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (eduardb) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: `Bitmap` is just how you hold in memory any kind of image, which doesn't matter on the file encoding. `BitmapFactory.decodeFile()` can decode both BMP and PNG, among others, without any extra effort required from your part :) Anyway, as you wish.

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (eduardb) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Would it work with less than 2 seconds? :D 
An alternative would be to maybe use a `WatchService` to see if the file has been modified, but not sure if it's worth it. Here is an article on how it would work: https://docs.oracle.com/javase/tutorial/essential/io/notification.html

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 1 above)
Comment (eduardb) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Fair enough. Do you think that it would make sense to have a parametrised test here though? To catch more edge cases.

File: app/src/main/java/net/aiscope/gdd_app/ui/util/BitmapReader.kt [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        cacheDir: File
+    ): Bitmap = withContext(Dispatchers.IO) {
+        val cachedImage = File(
+            cacheDir,
+            "${image.nameWithoutExtension}_${reqWidth}x${reqHeight}.${image.extension}"
+        )
+        if (cachedImage.exists()) {
+            return@withContext BitmapFactory.decodeFile(cachedImage.absolutePath)
+        }
+        val bitmap = BitmapReader.decodeSampledBitmapFromResource(
```
Comment (eduardb) [inline app/src/main/java/net/aiscope/gdd_app/ui/util/BitmapReader.kt:47]: `BitmapReader` is redundant here :D

**Review by eduardb — APPROVED** [review]
Looks solid to me. I am approving the PR, feel free to apply or not my suggestions from my last comments, and merge ;)

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Hahaha well, bigger than 1 second at least. 1200 ms would be acceptable I guess :P 

I don't think the `WatchService` the best idea, I was thinking more along the lines of using Mockito to see if `writeToFile` gets called twice. Not _exactly_ the same test but similar enough I guess.

File: app/src/main/java/net/aiscope/gdd_app/ui/util/BitmapReader.kt (hunk 7 above)
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/util/BitmapReader.kt:47]: Ah yeah. Fixed

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Hmmmm yeah it gets read from a JPG in most cases anyways. OK adjusted

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 1 above)
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Well these just test basic functionality. I am sure we _could_ add more tests for edge cases, but that sort of goes for the entire code base ;)

I don't think it's worth the effort right now

File: app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/androidTest/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapterTest.kt]: Changed it to `1200ms` for now
